library/alloc/src/vec/mod.rs

```diff
@@ -3183,6 +3183,57 @@ pub fn from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<
     <T as SpecFromElem>::from_elem(elem, n, alloc)
 }
 
+/// # Safety
+///
+/// `elem` must be the result of some const expression.
+#[doc(hidden)]
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "vec_of_const_expr", issue = "none")]
+#[track_caller]
+pub unsafe fn from_const_elem<T>(elem: T, n: usize) -> Vec<T> {
+    // SAFETY: Caller has guaranteed `elem` being the result of some const expression.
+    unsafe { from_const_elem_in(elem, n, Global) }
+}
+
+/// # Safety
+///
+/// `elem` must be the result of some const expression.
+#[doc(hidden)]
+#[cfg(not(no_global_oom_handling))]
+#[unstable(feature = "vec_of_const_expr", issue = "none")]
+#[track_caller]
+unsafe fn from_const_elem_in<T, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<T, A> {
+    /// # Safety
+    ///
+    /// `value` must points to a valid `T` value that is the result of some const expression.
+    unsafe fn fill_const_value<T>(buffer: &mut [MaybeUninit<T>], value: *const T) {
+        for target in buffer {
+            // SAFETY: If `value` is the result of some const expression, we can make as many
+            // copies as needed.
```

Comment:
This is an interesting argument that I have not seen before. I can't see anything immediately wrong with it, but paging in @rust-lang/opsem for extra scrutiny.

Reply:
This should be the case with the current `const` capabilities, but I'm not 100% sure it holds with every possible extension, namely leaking const allocation to runtime. I think the rules around const object unification would make this a valid execution, but I don't have all the relevant info in cache.

Reply:
We do intent that consts can still be implemented as many copies of the same thing for codegen. So whatever we do with const allocation should support that.

But it may be good to put a comment in the source here explaining that this is the standard library doing privileged reasoning, and not a stable guarantee.

Reply:
I have updated the safety comment. Please let me know if there is anything missing or inaccurate.